arch/x86/kernel/slmodule.c

```diff
@@ -147,104 +140,106 @@ static const struct file_operations sl_evtlog_ops = {
-					fs_entries[parent], NULL, ops);
-	if (IS_ERR(fs_entries[entry])) {
-		pr_err("Error creating securityfs %s file\n", name);
-		return PTR_ERR(fs_entries[entry]);
-	}
-	return 0;
-}
+/* sysfs file handles */
+static struct dentry *slaunch_dir;
+static struct dentry *event_file;
+static struct dentry *txt_dir;
+static struct dentry *txt_entries[SL_TXT_ENTRY_COUNT];
 
 static long slaunch_expose_securityfs(void)
 {
 	long ret = 0;
-	int i = 0;
+	int i;
 
-	fs_entries[SL_ROOT_DIR_ENTRY] = securityfs_create_dir("slaunch", NULL);
-	if (IS_ERR(fs_entries[SL_ROOT_DIR_ENTRY])) {
-		pr_err("Error creating securityfs slaunch root directory\n");
-		ret = PTR_ERR(fs_entries[SL_ROOT_DIR_ENTRY]);
-		goto err;
-	}
+	slaunch_dir = securityfs_create_dir("slaunch", NULL);
+	if (IS_ERR(slaunch_dir))
+		return PTR_ERR(slaunch_dir);
 
 	if (slaunch_get_flags() & SL_FLAG_ARCH_TXT) {
-		fs_entries[SL_TXT_DIR_ENTRY] =
-			securityfs_create_dir("txt",
-					      fs_entries[SL_ROOT_DIR_ENTRY]);
-		if (IS_ERR(fs_entries[SL_TXT_DIR_ENTRY])) {
-			pr_err("Error creating securityfs txt directory\n");
-			ret = PTR_ERR(fs_entries[SL_TXT_DIR_ENTRY]);
-			goto err_dir;
+		txt_dir = securityfs_create_dir("txt", slaunch_dir);
+		if (IS_ERR(txt_dir)) {
+			ret = PTR_ERR(txt_dir);
+			goto remove_slaunch;
 		}
 
-		for (i = 0; i < SL_TXT_ENTRY_COUNT; i++) {
-			ret = sl_create_file(SL_TXT_FILE_FIRST - i,
-					     sl_files[i].parent, sl_files[i].name,
-					     sl_files[i].fops);
-			if (ret)
-				goto err_dir;
+		for (i = 0; i < ARRAY_SIZE(sl_files); i++) {
+			txt_entries[i] = securityfs_create_file(
+						sl_txt_files[i].name, 0440,
+						txt_dir, NULL,
+						sl_txt_files[i].ops);
+			if (IS_ERR(txt_entries[i])) {
+				ret = PTR_ERR(txt_entries[i]);
+				goto remove_files;
+			}
 		}
+
 	}
 
 	if (sl_evtlog.addr > 0) {
-		ret = sl_create_file(0, SL_ROOT_DIR_ENTRY, sl_evtlog.name,
-				     &sl_evtlog_ops);
-		if (ret)
-			goto err_dir;
+		event_file = securityfs_create_file(
+					sl_evtlog.name, 0440,
+					slaunch_dir, NULL,
+					&sl_evtlog_ops);
+		if (IS_ERR(event_file)) {
+			ret = PTR_ERR(event_file);
+			goto remove_files;
+		}
 	}
 
 	return 0;
 
-err_dir:
-	for (i = 0; i <= SL_ROOT_DIR_ENTRY; i++)
-		securityfs_remove(fs_entries[i]);
-err:
+remove_files:
+	if (slaunch_get_flags() & SL_FLAG_ARCH_TXT) {
+		while (--i >= 0)
+			securityfs_remove(txt_entries[i]);
+		securityfs_remove(txt_dir);
+	}
+remove_slaunch:
+	securityfs_remove(slaunch_dir);
+
 	return ret;
 }
 
 static void slaunch_teardown_securityfs(void)
 {
 	int i;
 
-	for (i = 0; i < SL_FS_ENTRIES; i++)
-		securityfs_remove(fs_entries[i]);
+	securityfs_remove(event_file);
+	if (sl_evtlog.addr) {
+		memunmap(sl_evtlog.addr);
+		sl_evtlog.addr = NULL;
+	}
+	sl_evtlog.size = 0;
 
 	if (slaunch_get_flags() & SL_FLAG_ARCH_TXT) {
-		if (sl_evtlog.addr) {
-			memunmap(sl_evtlog.addr);
-			sl_evtlog.addr = NULL;
-		}
-		sl_evtlog.size = 0;
+		for (i = 0; i < SL_TXT_ENTRY_COUNT; i++)
```

Comment:
Above you use ARRAY_SIZE to get the upper bound while adding files. Would it be more consistent to do it the same way here?

Reply:
The justification to myself is that checking size of sl_txt_files before use makes sense above, but now in teardown sl_txt_files is not referenced, so to me it would be better to use the entry count. I am not set on that and can switch to ARRAY_SIZE here as well.

Reply:
Nope that is fine with me. Merging soon.